library/std/src/io/impls.rs

```diff
@@ -410,3 +411,40 @@ impl<A: Allocator> Write for Vec<u8, A> {
         Ok(())
     }
 }
+
+/// Read is implemented for `VecDeque<u8>` by consuming bytes from the front of the `VecDeque`.
+#[stable(feature = "vecdeque_read_write", since = "1.61.0")]
+impl<A: Allocator> Read for VecDeque<u8, A> {
+    #[inline]
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
```

Comment:
pondering: since `read` doesn't have to return everything at once, maybe take advantage of that?

This could be something like
```rust
    let source = self.as_slices().0;
    let n = source.len().min(buf.len());
    buf[..n].copy_from_slice(&source[..n]);
    self.drain(..n); // I wish there was an `rtruncate`
    Ok(n)
```

That way the `read` is always just a single `memcpy`.  And `read_exact` will get the rest of the data in the second `read` call, if needed, since `as_slices` always returns the empty slice as `.1`, if it happens to be contiguous.

(I'm not confident that LLVM's optimizer can split the zip-drain into two `memcpy`s.  And sure, you could write the one-or-two `memcpy`s yourself, but since the caller can't rely on it anyway, might as well make it `read_exact`'s problem.  Arguably it even makes it a *better* mock to sometimes do incomplete `read`s...)

Reply:
I like this proposal, and have it implemented in the latest commit! Thanks for the review.

I've also added similar logic to Write, such that `write` only expands the allocation of the VecDeque if there is no freespace at all. Write will favor a shorter transfer if it can prevent a reallocation.

Reply:
@scottmcm thanks for the review! I think I captured the idea, and in my latest version, it's much simpler of an implementation. I'll mark the thread as resolved, but please open if you had a different vision.